engine/launcher/src/main/scala/org/enso/launcher/installation/DistributionInstaller.scala

```diff
@@ -0,0 +1,344 @@
+package org.enso.launcher.installation
+
+import java.nio.file.Files
+
+import org.enso.cli.CLIOutput
+import org.enso.launcher.FileSystem.PathSyntax
+import org.enso.launcher.internal.OS
+import org.enso.launcher.internal.installation.DistributionManager
+import org.enso.launcher.{
+  FileSystem,
+  GlobalConfigurationManager,
+  InternalOpts,
+  Logger
+}
+
+import scala.util.control.NonFatal
+
+class DistributionInstaller(
+  manager: DistributionManager,
+  autoConfirm: Boolean,
+  bundleActionOption: Option[DistributionInstaller.BundleAction]
+) {
+  final private val installed = manager.LocallyInstalledDirectories
+  private val env             = manager.env
+
+  /**
+    * Names of additional files that are not essential to running the
+    * distribution, but should be copied over to the data root if possible.
+    * These files are assumed to be located at the data root.
+    */
+  private val nonEssentialFiles       = Seq("README.md", "NOTICE")
+  private val nonEssentialDirectories = Seq("components-licences")
+
+  private val enginesDirectory =
+    installed.dataDirectory / manager.ENGINES_DIRECTORY
+  private val runtimesDirectory =
+    installed.dataDirectory / manager.RUNTIMES_DIRECTORY
+
+  /**
+    * Installs the distribution under configured location.
+    * Unless [[autoConfirm]] is true, asks the user to confirm the action after
+    * printing where it plans to install itself.
+    */
+  def install(): Unit = {
+    try {
+      prepare()
+      installBinary()
+      createDirectoryStructure()
+      installBundles()
+      Logger.info("Installation succeeded.")
+      maybeRemoveInstaller()
+    } catch {
+      case NonFatal(e) =>
+        val message = s"Installation failed with error: $e."
+        Logger.error(message)
+        CLIOutput.println(message)
+        e.printStackTrace()
+        sys.exit(1)
+    }
+
+  }
+
+  /**
+    * Prepares for the installation - find and report possible conflicts, and
+    * ask the user if they want to proceed (unless [[autoConfirm]] is set, in
+    * which case it only reports conflicts).
+    */
+  private def prepare(): Unit = {
+    if (Files.exists(installed.dataDirectory)) {
+      Logger.warn(s"${installed.dataDirectory} already exists.")
+      if (!Files.isDirectory(installed.dataDirectory)) {
```

Comment:
Maybe have a flag to ignore this issue? If I want to upgrade, just remove it. Maybe this could be a question?

Reply:
If the directory just exists, only a warning is printed and the installer happily continues. So you can upgrade (although the standard way to upgrade will use `enso upgrade` anyway).

The error is thrown if that is not a directory but some kind of file. This should never happen if the user installed things normally. I'm not sure if overwriting the file without asking is something the users would expect.

Reply:
ok